src/lib/x509/x509_ext.cpp

```diff
@@ -847,6 +856,518 @@ void TNAuthList::decode_inner(const std::vector<uint8_t>& in) {
+
+      // 5 instead of 4 because there might be a leading 0 in some cases, e.g. 2^32-1
+      if(bytes.empty() || bytes.size() > 5) {
+         throw Decoding_Error("AS Range numbers must not be empty and have a length of up to 32 bit.");
+      }
+
+      // if we do have 5 octets, the first one is always a 0 so it would have no influence anyway
+      if(bytes.size() == 5) {
+         bytes.erase(bytes.begin());
+      }
+
+      for(size_t i = 0; i < bytes.size(); i++) {
+         m_min |= (static_cast<asnum_t>(bytes[i])) << 8 * (bytes.size() - 1 - i);
+      }
+
+      BER_Object as_max_obj = obj_ber.get_next_object();
+      bytes.assign(as_max_obj.data().begin(), as_max_obj.data().end());
+      if(bytes.empty() || bytes.size() > 5) {
+         throw Decoding_Error("AS Range numbers must not be empty and have a length of up to 32 bit.");
+      }
+
+      if(bytes.size() == 5) {
+         bytes.erase(bytes.begin());
+      }
+
+      for(size_t i = 0; i < bytes.size(); i++) {
+         m_max |= (static_cast<asnum_t>(bytes[i])) << 8 * (bytes.size() - 1 - i);
+      }
+
+      if(m_min > m_max) {
+         throw Decoding_Error("AS Range numbers must be sorted");
+      }
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for ASIdOrRange {}", static_cast<uint32_t>(type_tag)));
+   }
+}
+
+std::vector<uint8_t> IPAddressBlocks::encode_inner() const {
+   std::vector<uint8_t> output;
+   DER_Encoder(output).start_sequence().encode_list(this->m_ip_addr_blocks).end_cons();
+   return output;
+}
+
+void IPAddressBlocks::decode_inner(const std::vector<uint8_t>& in) {
+   BER_Decoder(in).decode_list(m_ip_addr_blocks).verify_end();
+}
+
+void IPAddressBlocks::IPAddressFamily::encode_into(Botan::DER_Encoder& into) const {
+   into.start_sequence();
+   into.add_object(ASN1_Type::OctetString, ASN1_Class::Universal, this->m_addr_family);
+   if(std::holds_alternative<IPAddressChoice<Version::IPv4>>(this->m_ip_addr_choice)) {
+      auto ipv4_choice = std::get<IPAddressChoice<Version::IPv4>>(this->m_ip_addr_choice);
+      into.encode(ipv4_choice);
+   } else {
+      auto ipv6_choice = std::get<IPAddressChoice<Version::IPv6>>(this->m_ip_addr_choice);
+      into.encode(ipv6_choice);
+   }
+   into.end_cons();
+}
+
+void IPAddressBlocks::IPAddressFamily::decode_from(Botan::BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
+   if(type_tag != ASN1_Type::Sequence) {
+      throw Decoding_Error(fmt("Unexpected type for IPAddressFamily {}", static_cast<uint32_t>(type_tag)));
+   }
+   BER_Decoder obj_ber = BER_Decoder(obj);
+
+   std::vector<uint8_t> addr_family;
+   obj_ber.decode(addr_family, ASN1_Type::OctetString);
+   m_addr_family = addr_family;
+   size_t addr_family_length = addr_family.size();
+   if(2 > addr_family_length || addr_family_length > 3) {
+      throw Decoding_Error("AFI/SAFI too long / too short.");
```

Comment:
I find this surpsisingly hard to parse. Maybe just

`if(addr_family.size() != 2 && addr_family.size() != 3)`

?